### PR TITLE
multiple input argument version of adj_jac_apply (Issue #963)

### PR DESCRIPTION
**Pull request by bbbales2**
## Summary

This is basically the C++ version of the stuff described in issue #963 taking advantage of C++ parameter packs and recursively templated functions to make it all work except that int or std::vector<double>, std::vector<Eigen::Matrix>, and std::vector<int> arguments are not supported yet.

There are three recursively templated member functions of adj_jac_vari at work here. count_memory, which goes through the argument list and figures out how many varis there are, prepare_x_vis, which actually copies the vari pointers out of the input vars, and accumulate_adjoints, which takes the results of the user-supplied adjoint_jacobian_multiply and actually accumulates the input adjoints.

There's a lot of recursively templated functions. I chose to just document the functionality of them all together right before defining them, rather than documenting them each individually. There's kindof a blowup in the number of functions you gotta define to make this correct. In C++17 that'll reduce to two so documenting won't be such a hassle, but for now I've just avoided putting docs everywhere cause I think it'd be a bit too much.

I used a C++14 replacement for C++17's apply operator (https://github.com/stan-dev/math/blob/44e0915f30bd21adf1743506ad0e3d85896bb3cb/stan/math/rev/mat/fun/adj_jac_apply.hpp#L20) that I took from http://aherrmann.github.io/programming/2016/02/28/unpacking-tuples-in-cpp14/ .

## Tests

Run tests with:
```./runTests.py test/unit/math/rev/mat/fun/adj_jac_apply_test```

I added three new adj_jac_vari functors to test/unit/math/rev/mat/fun/adj_jac_apply_test, one that takes in two Eigen::Matrix<double, Eigen::Dynamic, 1> arguments, one that takes one Eigen::Matrix<double, Eigen::Dynamic, 1> as the first argument and a scalar as a second, and one that takes a scalar as the first argument and an Eigen::Matrix<double, Eigen::Dynamic, 1> as the second.

The autodiff is tested with both arguments being vars, and then the mixes of one being a var and one a double.

## Side Effects

Updated softmax to work with new adj_jac_apply

## Checklist

- [x] Math issue #963 

- [x] Copyright holder: Columbia University

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by bbbales2** [comment]
@bob-carpenter This is ready to review

**Comment by wds15** [comment]
ui... this looks like some very cool new C++ stuff. Nice.

**Comment by bob-carpenter** [comment]
I'll get to it ASAP.  Probably tomorrow (Tuesday).

**Comment by bbbales2** [comment]
@bob-carpenter I made the changes that we talked about yesterday (cleaned up the recursions n' such). I also added in support for int, std::vector<int>, and std::vector<double> argument types. I think I'm happy with that.

The code itself should be ready to review (if the tests pass), but the current tests are inadequate. You can review the code now if you want and skip the tests until I finish, or if I hear nothing else I'll ping you when I finish the tests and you can do it all in one big pass.

**Comment by bbbales2** [comment]
> Just make sure to put the std::vector definitions after the matrix ones so you can instantiate vectors of matrices (if the recursions work that way).

> If we could carry this all the way through to vectorization

I was thinking about that a little but I decided to pump the breaks on std::vectors of Eigen types. I think a vectorized version of this needs more thought, and might benefit more from a different type of functor. I ddin't want to pre-emptively introduce any more requirements on the functors we have cause I think they look nice.

I think the prototype for this sorta thing though should be the GPs, and @drezap has done quite a few at this point so we've got a lot of material to look at when we get to that point.

I'll letcha know when the tests are in place!

**Comment by bbbales2** [comment]
@bob-carpenter This is ready to review (once the tests pass)

**Comment by bbbales2** [comment]
@bob-carpenter Tests passed! Review away

**Comment by wds15** [comment]
If I am getting it right what you are doing here then this a super cool replacement for `operands_and_partials`.... and this approach should have much better memory locality (since all is done once in a single sweep) and potentially we can even parallelize this stuff more efficiently (as calculation of the partials and summing them up is done in a single go). Am I right in this?

If so... does it make sense to look at the code once more for opportunities to parallelize things? I don't know at this stage if there is anything smart one can do / you may just want to continue and do this (if at all) later / it is not needed.

Anyway, I will try this PR out on some lpmfs to see if I get speedups for large models. I am almost sure I will. If so (which I expect), then this is a killer feature!

**Comment by bob-carpenter** [comment]
I don't think you get extra memory locality or should expect a lot more speed from adj_jac_apply than you'd get from operands_and_partials.  It was mainly designed for multivariate output and the _lpdf, _lpmf, etc. only have a single output.  With a single output, storing the Jacobian is still linear in the input---with multivariate output it's M * N.

The other big advantage is that it should be more natural for applied math and stats people to write the adjoint-Jacobian product than to implement something with operands_and_partials.

I think we'll still need the expression template views of scalars/vectors and still need to be careful about where we loop and do extra computations.

For localization, the problem is the inputs.  There's no way to guarantee locality there.  The output's just a single vari here, so there's no difference in locality there.

**Comment by wds15** [comment]
Having gone through implementing the Poisson lpmf I agree in that there is no benefit in terms of locality, etc.. I still got a 7.5% speedup when using the adjoint facility though in comparison to using the vanilla one. This could be due to less checking, though.

Anyway, this is nice stuff!

**Comment by bbbales2** [comment]
> If I am getting it right what you are doing here then this a super cool replacement for operands_and_partials

The biggest deficiency compared to operands_and_partials is this is exclusively reverse mode. I think operands_and_partials covers everything. Not sure, but it at least does reverse and forwards.

But this being specialized to reverse mode means we aren't having to work with the full Jacobians! But like Bob said the lpdfs/lpmfs all only have one outputs, so the advantage there is moot.

A lot of the speedup here is gonna come from only having one varis on the chaining autodiff stack with vector/matrix output functions.

> I still got a 7.5% speedup when using the adjoint

Might not be the biggest speedup, but I am glad to hear this! I've keep telling myself and looking at the code and saying that this is as efficient as doing thing the other way, but I hadn't actually checked lol.

> If so... does it make sense to look at the code once more for opportunities to parallelize things?

Hmm, that's a good thought I think. One thing this does do is (mostly) isolate the autodiff stack from the implementation of the function + adjoint/Jacobian. At least it's isolating the adjoint accumulation and output vari allocation n' such.

Something I've been avoiding on this pull req. are vectorization-style types. Like `std::vector<Eigen::VectorXd>` and stuff. Maybe it'd make sense to make a vectorizing version of this thing that could also take advantage of various parallelization backends?

Like, maybe the user implements:

```
struct Functor {
  VectorXd operator()(const VectorXd& arg1, const VectorXd& arg2);
  auto chain(VectorXd& adj);
}
```

And then there's an adj_jac_parallel_apply that wraps this at a higher level and exposes at the interface:

```
func(VectorXd, VectorXd)
func(std::vector<VectorXd>, VectorXd)
func(VectorXd, std::vector<VectorXd>)
func(std::vector<VectorXd>, std::vector<VectorXd>)
```

And broadcasts things in the usual Stan lpdf way.

Since the user op isn't touching the autodiff stack, you could schedule and make the individual calls and accumulate the results as you pleased.

I dunno if this would do any good. Seems like it could be cool.

You could do some weird things. Like maybe if operator() and chain can all get called out of order and in parallel, you guarantee another function (initialize or whatever) gets called once before all the rest, and that thread can do any calculations that will be shared.

I dunno for sure, but you might be on to something there. This would basically be automatically parallelizing ops that are put in to adj_jac_apply form.

**Comment by bbbales2** [comment]
@wds15 also to be clear I hope to have this with different return types out within a couple weeks. If you did like the idea of making an automatically vectorizing version, that's the timeframe to where you'd have stable code to work from.

File: stan/math/rev/mat/fun/adj_jac_apply.hpp [hunk 1]
```diff
@@ -18,34 +35,224 @@ namespace math {
...
+  }
+
+  template <typename... Pargs>
+  void prepare_x_vis(const var& x, const Pargs&... args) {
+    constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
+    x_vis_[offsets_[t]] = x.vi_;
+    prepare_x_vis(args...);
+  }
+
+  template <typename... Pargs>
```
Comment (bob-carpenter) [inline stan/math/rev/mat/fun/adj_jac_apply.hpp]: [optional]
I would think both the int/double here and above could be replaced with a single template.

File: stan/math/rev/mat/fun/adj_jac_apply.hpp [hunk 2]
```diff
@@ -18,34 +35,224 @@ namespace math {
...
+  void prepare_x_vis(const std::vector<int>& x, const Pargs&... args) {
+    prepare_x_vis(args...);
+  }
+
+  template <typename... Pargs>
+  void prepare_x_vis(const var& x, const Pargs&... args) {
+    constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
```
Comment (bob-carpenter) [inline stan/math/rev/mat/fun/adj_jac_apply.hpp]: This is fiendishly clever, by the way (computing sizes this way).  I do worry a bit about compile times.

File: stan/math/rev/mat/fun/adj_jac_apply.hpp [hunk 3]
```diff
@@ -18,34 +35,224 @@ namespace math {
...
    */
-  explicit adj_jac_vari(const Eigen::Matrix<var, Eigen::Dynamic, 1>& x)
+  explicit adj_jac_vari(const Targs&... args)
       : vari(std::numeric_limits<double>::quiet_NaN()),  // The val_ in this
                                                          // vari is unused
-        N_(x.size()),
-        x_vi_(build_vari_array(x)) {
-    Eigen::Matrix<double, Eigen::Dynamic, 1> val_x = value_of(x);
-    Eigen::Matrix<double, Eigen::Dynamic, 1> val_y = f_(val_x);
+        is_var_({{is_var<typename scalar_type<Targs>::type>::value...}}),
```
Comment (bob-carpenter) [inline stan/math/rev/mat/fun/adj_jac_apply.hpp]: I still can't get over that these two lines will work.

File: stan/math/rev/mat/fun/adj_jac_apply.hpp [hunk 4]
```diff
@@ -54,9 +261,88 @@ struct adj_jac_vari : public vari {
...
+   * If a1 is an Eigen::Matrix<double, 5, 5>, a2 is a double, a3 is a double,
+   * a4 is a length 5 std::vector of doubles, a5 is a length 3 std::vector of
+   * ints and is_var_ is set to {true, false, true, true, false}, the 25 values
+   * in a1 are accumulated in the adjs of the varis pointed to by the values of
+   * x_vis_ to x_vis_ + 24, the value of a2 is ignored, and the adjoint of the
+   * vari pointed to by x_vis_[25] is incremented by a3, the 5 values in a4 are
+   * accumulated in the adjs of the varis pointed to by the values of x_vis_ +
+   * 26 to x_vis_ + 30, and the values in a5 are ignored.
+   *
+   * @param y_adj_jac next set of adjoints to be accumulated
```
Comment (bob-carpenter) [inline stan/math/rev/mat/fun/adj_jac_apply.hpp]: Each of these functions should get doc including tparam.  

I can see why you want to skip it, as they're all just part of the same definition, but not from the C++ perspective where each one is completely independent as they're just overloads.

File: stan/math/rev/mat/fun/adj_jac_apply.hpp [hunk 5]
```diff
@@ -66,10 +352,15 @@ struct adj_jac_vari : public vari {
...
-        = f_.multiply_adjoint_jacobian(y_adj);
-    for (int n = 0; n < N_; ++n)
-      x_vi_[n]->adj_ += y_adj_jac(n);
+    auto y_adj_jacs = f_.multiply_adjoint_jacobian(is_var_, y_adj);
+
+    /**
+     * NOTE TO REVIEWER: If the cost of this lambda function redirection seems
+     * to high, we could make accumulate_adjoints a non member function and just
+     * pass it all the info it needs.
+     */
```
Comment (bob-carpenter) [inline stan/math/rev/mat/fun/adj_jac_apply.hpp]: The best way to figure out the cost is to test it.  I'm not sure how much overhead is involved where with the lambdas.  If they don't bind anything, they're simple, but here it's binding `this`.  I'd have thought you'd have wanted to bind it by reference, but I guess it's actually just a pointer until it's used as `*this` or `this->...`.

File: stan/math/rev/mat/fun/adj_jac_apply.hpp [hunk 6]
```diff
@@ -120,14 +429,18 @@ struct adj_jac_vari : public vari {
...
+ * @param args input to the functor
  * @return the result of the specified operation wrapped up in vars
  */
-template <class F>
-Eigen::Matrix<var, Eigen::Dynamic, 1> adj_jac_apply(
-    const Eigen::Matrix<var, Eigen::Dynamic, 1>& x) {
-  adj_jac_vari<F>* vi = new adj_jac_vari<F>(x);
+template <typename F, typename... Targs>
+Eigen::Matrix<var, Eigen::Dynamic, 1> adj_jac_apply(const Targs&... args) {
+  auto vi = new adj_jac_vari<F, Targs...>(args...);
```
Comment (bob-carpenter) [inline stan/math/rev/mat/fun/adj_jac_apply.hpp]: This constructor for `adj_jac_vari` is magical.

File: test/unit/math/rev/mat/fun/adj_jac_apply_test.cpp [hunk 7]
```diff
@@ -1,12 +1,1198 @@
...
 #include <sstream>
+#include <tuple>
+#include <vector>
 
+/**
+ * Test that all types that are documented to be supported can actually be
+ * included in functor. Check initialization of is_vars_ and offsets_ are to
+ * specification.
+ */
+struct WeirdArgumentListFunctor1 {
```
Comment (bob-carpenter) [inline test/unit/math/rev/mat/fun/adj_jac_apply_test.cpp]: [optional]
This doesn't need to be rewritten, but going forward, more smaller tests are a lot easier to deal with and a lot easier to use to diagnose problems.

File: test/unit/math/rev/mat/fun/adj_jac_apply_test.cpp [hunk 8]
```diff
@@ -1,12 +1,1198 @@
...
+  y1(0).grad();
+
+  auto vi1 = make_vari_for_test<WeirdArgumentListFunctor1>(
+      d, i, d, i, vd, vi, vd, vi, ed1, ed2, ed3, ed4, ed1, ed2, ed3, ed4);
+
+  EXPECT_TRUE(
+      (vi1->is_var_
+       == std::array<bool, 16>(
+              {{false, false, false, false, false, false, false, false, false,
+                false, false, false, false, false, false, false}})));
```
Comment (bob-carpenter) [inline test/unit/math/rev/mat/fun/adj_jac_apply_test.cpp]: [comment]
These long lists are a symptom of the problem with fat tests.

File: test/unit/math/rev/mat/fun/adj_jac_apply_test.cpp [hunk 9]
```diff
@@ -1,12 +1,1198 @@
...
+
+  auto vi2 = make_vari_for_test<WeirdArgumentListFunctor1>(
+      v, i, d, i, vv, vi, vd, vi, ev1, ed2, ev3, ed4, ev1, ed2, ev3, ed4);
+
+  EXPECT_TRUE((vi2->is_var_
+               == std::array<bool, 16>(
+                      {{true, false, false, false, true, false, false, false,
+                        true, false, true, false, true, false, true, false}})));
+
+  EXPECT_TRUE((vi2->offsets_
```
Comment (bob-carpenter) [inline test/unit/math/rev/mat/fun/adj_jac_apply_test.cpp]: I think this should be `EXPECT_EQ` here and elsewhere.

File: test/unit/math/rev/mat/fun/adj_jac_apply_test.cpp [hunk 10]
```diff
@@ -1,12 +1,1198 @@
...
+    ed3(i) = 1.0;
+
+  ed4(0) = 1.0;
+
+  Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1> y
+      = stan::math::adj_jac_apply<CheckAdjointsPassingThrough>(
+          d, vd, 5, ed1, ed2, vi, ed3, ed4);
+
+  y(0).grad();
+  EXPECT_NEAR(y(0).val(), d.val(), 1e-10);
```
Comment (bob-carpenter) [inline test/unit/math/rev/mat/fun/adj_jac_apply_test.cpp]: Unless the tests don't pass, these `EXPECT_NEAR` should be replaced with `EXPECT_FLOAT_EQ`.

File: test/unit/math/rev/mat/fun/adj_jac_apply_test.cpp [hunk 11]
```diff
@@ -1,12 +1,1198 @@
...
+    y(1 + j).grad();
+    EXPECT_NEAR(y(1 + j).val(), vd[j].val(), 1e-10);
+    EXPECT_NEAR(vd[j].adj(), 1.0, 1e-10);
+    EXPECT_NEAR(d.adj(), 0.0, 1e-10);
+    for (int i = 0; i < size_vd; i++)
+      if (i != j)
+        EXPECT_NEAR(vd[i].adj(), 0, 1e-10);
+    for (int i = 0; i < rows_ed1; i++)
+      EXPECT_NEAR(ed1(i).adj(), 0, 1e-10);
+    for (int i = 0; i < rows_ed2 * cols_ed2; i++)
```
Comment (bob-carpenter) [inline test/unit/math/rev/mat/fun/adj_jac_apply_test.cpp]: Overall, I'd rather be testing with finite diffs than with hand-written derivatives like this.

File: test/unit/math/rev/mat/fun/adj_jac_apply_test.cpp (hunk 7 above)
Comment (bbbales2) [reply test/unit/math/rev/mat/fun/adj_jac_apply_test.cpp]: Yeah, I agree. I made it big and bulky cause I thought that'd be more likely to cause a breakdown in the components though. But I guess that's a double-edged technique.

File: stan/math/rev/mat/fun/adj_jac_apply.hpp (hunk 5 above)
Comment (bbbales2) [reply stan/math/rev/mat/fun/adj_jac_apply.hpp]: I went to Google how to bind by reference, and found out you cannot bind "this" by reference: https://stackoverflow.com/questions/16323032/why-cant-i-capture-this-by-reference-this-in-lambda . Who knew?

File: stan/math/rev/mat/fun/adj_jac_apply.hpp (hunk 3 above)
Comment (bbbales2) [reply stan/math/rev/mat/fun/adj_jac_apply.hpp]: I didn't know what C++17 folds were and why I should care until I was playing with the include_summands stuff in the @dpsimpson thread:

```
value = (!propto || (!stan::is_constant<typename scalar_type<T_pack>::type>::value || ...))
```

Is fine hahaha.

File: stan/math/rev/mat/fun/adj_jac_apply.hpp (hunk 1 above)
Comment (bbbales2) [reply stan/math/rev/mat/fun/adj_jac_apply.hpp]: I made the templates here explicit cause I wanted it to fail if someone tried to pass a type through I didn't recognize. This seems like the type of class where folks might try to pass weird things in and expect something that isn't implemented.

File: stan/math/prim/scal/fun/value_of.hpp [hunk 12]
```diff
@@ -40,6 +40,19 @@ inline double value_of<double>(double x) {
...
+ *
+ * <p>See <code>value_of(T)</code> for a polymorphic
+ * implementation using static casts.
+ *
+ * <p>This inline pass-through no-op should be compiled away.
+ *
+ * @param x value
+ * @return input value
+ */
+inline int value_of(const int& x) { return x; }
```
[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/prim/scal/fun/value_of.hpp]: Yeah I don't know which would be preferred, so I just chose this for symmetry.

**Comment by bob-carpenter** [comment]
...

> I made the templates here explicit cause I wanted it to fail if someone tried to pass a type through I didn't recognize. This seems like the type of class where folks might try to pass weird things in and expect something that isn't implemented.

I wouldn't be too worried about that happening---it has to implement the concepts used, so it's not like you can just drop anything in.  But it's OK to leave it for now.

**Comment by bob-carpenter** [comment]
> Yeah I don't know which would be preferred, so I just chose this for symmetry.

For read-only arguments (those that don't pass in references or pointers that can get changed), pass primitives by value and classes by constant reference.

**Comment by bbbales2** [comment]
> For read-only arguments (those that don't pass in references or pointers that can get changed), pass primitives by value and classes by constant reference.

I made this change. This is ready to review when the tests pass.

**Comment by bbbales2** [comment]
WooooOoOOoooO @bob-carpenter @bob-carpenter @bob-carpenter look look it passed!

File: stan/math/rev/mat/fun/adj_jac_apply.hpp [hunk 13]
```diff
@@ -120,14 +457,18 @@ struct adj_jac_vari : public vari {
...
   Eigen::Matrix<var, Eigen::Dynamic, 1> y(vi->M_);
 
   for (int m = 0; m < y.size(); ++m)
```
Comment (bob-carpenter) [inline stan/math/rev/mat/fun/adj_jac_apply.hpp]: [optional]
[optional, but a really good idea]

Use `std::copy`:

```
std::copy(&y[0], &y[0] + y.size(), vi->y_vi_);
```

the reason to favor `std::copy` is efficiency:(this from cppreference):

> In practice, implementations of std::copy avoid multiple assignments and use bulk copy functions such as std::memmove if the value type is TriviallyCopyable

And it turns out our `var` are TriviallyCopyable:

```
  EXPECT_TRUE(std::is_trivially_copyable<stan::math::var>::value);
```

passes.

**Comment by bbbales2** [comment]
This should be good to go.

**Pull request merged**